legion/sdk/legion/sdk/clients/edi.py

```diff
@@ -161,15 +239,57 @@ def _request(self, url_template: str, payload: typing.Mapping[typing.Any, typing
         else:
             raise EDIConnectionException('Can not reach {}'.format(self._base)) from raised_exception
 
-        # We assume if there were redirects then credentials are out of date
+        # We assume if there were redirects then credentials are out of date and we can refresh or build auth url
         if response.history:
-            LOGGER.debug('Status code: "{}", Response: "{}"'.format(response.status_code, response.text))
+            # If it is a error after refreshed token - fail
+            if limit_stack:
+                raise IncorrectAuthorizationToken(
+                    'Credentials are not correct even after refreshing. \n'
+                    'Please login again'
+                ) from raised_exception
 
-            parse_result = urlparse(target_url)
-            raise IncorrectAuthorizationToken(
-                'Credentials are not correct. You should open {}://{} url in a browser to get fresh token'.format(
-                    parse_result.scheme, parse_result.netloc
-                )) from raised_exception
+            LOGGER.debug('Status code: "{}", Response: "{}"'.format(response.status_code, response.text))
+            LOGGER.debug('Redirect has been detected. Trying to refresh a token')
+
+            if legion.sdk.config.EDI_REFRESH_TOKEN and legion.sdk.config.EDI_ISSUING_URL:
+                LOGGER.debug('Refresh token for %s has been found, trying to use it', legion.sdk.config.EDI_ISSUING_URL)
+                login_result = do_refresh_token(legion.sdk.config.EDI_REFRESH_TOKEN, legion.sdk.config.EDI_ISSUING_URL)
+                if not login_result:
+                    raise IncorrectAuthorizationToken(
+                        'Refresh token in not correct. \n'
+                        'Please login again'
+                    ) from raised_exception
+                else:
+                    self._update_config_with_new_oauth_config(login_result)
+                    return self._request(
+                        url_template,
+                        payload=payload,
+                        action=action,
+                        stream=stream,
+                        timeout=timeout,
+                        limit_stack=True
+                    )
+            else:
+                if self._non_interactive:
+                    raise IncorrectAuthorizationToken(
+                        'Credentials are not correct. \n'
+                        'Please provide correct temporary token or disable non interactive mode'
+                    ) from raised_exception
+                else:
+                    # Start interactive flow
+                    self._interactive_login_finished.clear()
+                    target_url = get_authorization_redirect(response.url, self.after_login)
+                    LOGGER.error('Credentials are not correct. \nPlease open %s', target_url)
```

Comment:
When I try to login first time with an empty config, I get `Credentials are not correct` message, but I did not provide theirs yet. So this message can be confusing for users.

Reply:
Done